crates/db_schema/src/impls/password_reset_request.rs

```diff
@@ -64,22 +69,51 @@ impl PasswordResetRequest {
     let form = PasswordResetRequestForm {
       local_user_id: from_local_user_id,
       token_encrypted: token_hash,
+      expires_at: naive_now() + Duration::days(1),
     };
 
+    // TODO do this in a transaction along with the new token creation (blocked by https://github.com/LemmyNet/lemmy/issues/1161)
+    Self::expire_all_for_user(pool, from_local_user_id).await?;
+
     Self::create(pool, &form).await
   }
-  pub async fn read_from_token(pool: &DbPool, token: &str) -> Result<PasswordResetRequest, Error> {
+
+  pub async fn read_unexpired_from_token(
+    pool: &DbPool,
+    token: &str,
+  ) -> Result<PasswordResetRequest, Error> {
     let conn = &mut get_conn(pool).await?;
     let mut hasher = Sha256::new();
     hasher.update(token);
     let token_hash: String = bytes_to_hex(hasher.finalize().to_vec());
```

Comment:
I didn't write this, so I have no idea why its necessary. What's wrong with just emailing a UUID, and checking it? What do we gain by converting this uuid back and forth to hex?

Reply:
It's not just converting to hex, it's hashing it with SHA256. If an attacker gains read access to the database in some way that means they won't be able to reverse it in order to use it themselves.

Reply:
This isn't directed at you, but we should be using the bcrypt hash and verify functions, rather than rolling our own here.

Reply:
No, SHA256 is perfectly fine in this case since it's being used with a cryptographically random input. Bcrypt is designed to use a lot of iterations to make brute forcing low entropy passwords more difficult, but when you already have a high entropy input there's no need to waste cycles on bcrypt. Salting is unnecessary for the same reason, every token high entropy so they're effectively guaranteed to be  unique.

Reply:
Gotcha. It all seems fairly pointless to me since these tokens shouldn't last very long. And if an attacker has direct DB access they can already change a user's password to whatever they like.

Hashing passwords makes sense so that a DB leak can't reveal plaintext passwords. Hashing randomly generated short-lived UUIDs makes less sense.

Reply:
Yeah, it's not really a big deal to store them directly either since they're short lived, but it's already implemented this way and it's not _totally_ useless. It's also theoretically possible an attacker could get access to a recent database backup without being able to write to the live database, so there is _some_ threat, even if it's small.

Reply:
You are putting too much complexity for hypothetical scenarios that will probably never happen. Keep in mind that we will have to maintain this code for the next years. So get rid of the hash and use a simple uuid.

Reply:
@Nutomic It's _already_ using the hash, at this point it's more work to remove it and removing it would invalidate existing tokens. And while it's a hypothetical scenario with a small risk, I'd say it's a greater risk with federated applications like this since you have no control over everyone else's production environment and how they handle backups.